Assembly-CSharp/ModLoader.cs

```diff
@@ -488,36 +510,56 @@ IEnumerator PreloadScene(string s)
                         modScenePreloadedObjects[objName] = obj;
                     }
                 }
-
-                // Update loading progress
-                progress++;
-
-                updateLoadingBarProgress();
-                yield return USceneManager.UnloadSceneAsync(scene);
-                updateLoadingBarProgress();
             }
 
-            List<IEnumerator> batch = new();
-            int maxKeys = toPreload.Keys.Count;
-
-            foreach (string sceneName in toPreload.Keys)
+            void CleanupPreloadOperation(string sceneName)
             {
-                int batchCount = Math.Min(ModHooks.GlobalSettings.PreloadBatchSize, maxKeys);
-
-                batch.Add(PreloadScene(sceneName));
-
-                if (batch.Count < batchCount)
-                    continue;
+                Logger.APILogger.LogFine($"Unloading scene \"{sceneName}\"");
+                var unloadOperation = USceneManager.UnloadSceneAsync(sceneName);
+                unloadOperation.completed += (op) =>
+                {
+                    sceneProgress[sceneName] = 1.0f;
+                    preloadOperationQueue.Remove(unloadOperation);
+                };
+                preloadOperationQueue.Add(unloadOperation);
+            }
 
-                Coroutine[] coros = batch.Select(nb.StartCoroutine).ToArray();
+            void StartPreloadOperation(string sceneName)
+            {
+                Logger.APILogger.LogFine($"Loading scene \"{sceneName}\"");
+                sceneProgress[sceneName] = 0.0f;
+                var preloadOperation = USceneManager.LoadSceneAsync(sceneName, LoadSceneMode.Additive);
+                preloadOperation.priority = scenePriority[sceneName];
+                preloadOperation.completed += (op) =>
+                {
+                    sceneProgress[sceneName] = 0.5f;
+                    preloadOperationQueue.Remove(preloadOperation);
+                    GetPreloadObjectsOperation(sceneName);
+                    CleanupPreloadOperation(sceneName);
+                };
+                preloadOperationQueue.Add(preloadOperation);
+            }
 
-                foreach (var coro in coros)
-                    yield return coro;
+            int i = 0;
+            while (sceneProgress.Values.Average() < 1.0)
```

Comment:
This would be better put as just checking the completion status on the async operations themselves, you can then queue up more as-needed. This would also remove the need for callback style code.